quantus/helpers/perturb_func.py

```diff
@@ -8,134 +8,160 @@
+    arr: np.array,
+    indices: Tuple[np.array],
+    indexed_axes: Sequence[int],
+    blur_kernel_size: Union[int, Sequence[int]] = 15,
+    **kwargs,
 ) -> np.array:
     """
-    Replace a single patch in an array by a blurred version.
-    Blur is performed via a 2D convolution.
+    Replace array at indices by a blurred version.
+    Blur is performed via convolution.
     blur_kernel_size controls the kernel-size of that convolution (Default is 15).
-    Assumes unbatched channel first format.
     """
-    nr_channels = arr.shape[0]
-    # Create blurred array.
-    blur_kernel_size = (1, *([blur_kernel_size] * (arr.ndim - 1)))
+
+    indices = expand_indices(arr, indices, indexed_axes)
+
+    # Expand blur_kernel_size
+    if isinstance(blur_kernel_size, int):
+        blur_kernel_size = [blur_kernel_size for _ in indexed_axes]
+
+    assert len(blur_kernel_size) == len(indexed_axes)
+
+    # Create kernel and expand dimensions to arr.ndim
     kernel = np.ones(blur_kernel_size, dtype=arr.dtype)
     kernel *= 1.0 / np.prod(blur_kernel_size)
-    kernel = np.tile(kernel, (nr_channels, 1, *([1] * (arr.ndim - 1))))
-
-    if arr.ndim == 3:
-        arr_avg = conv2D_numpy(
-            x=arr,
-            kernel=kernel,
-            stride=1,
-            padding=0,
-            groups=nr_channels,
-            pad_output=True,
-        )
-    elif arr.ndim == 2:
-        raise NotImplementedError("1d support not implemented yet")
-    else:
-        raise ValueError("Blur supports only 2d inputs")
 
-    # Perturb array.
+    # Compute blur array. It is only blurred at indices, otherwise it is equal to arr.
+    # We only blur at indices since otherwise n-d convolution can be quite computationally expensive
+    arr_perturbed = blur_at_indices(arr, kernel, indices, indexed_axes)
+
+    return arr_perturbed
+
+
+def gaussian_noise(
+    arr: np.array,
+    indices: Union[int, Sequence[int], Tuple[np.array]],
+    indexed_axes: Sequence[int],
+    perturb_mean: float = 0.0,
+    perturb_std: float = 0.01,
+    **kwargs,
+) -> np.array:
+    """Add gaussian noise to the input."""
+
+    indices = expand_indices(arr, indices, indexed_axes)
+    noise = np.random.normal(loc=perturb_mean, scale=perturb_std, size=arr.shape)
+
     arr_perturbed = copy.copy(arr)
-    arr_perturbed[patch_slice] = arr_avg[patch_slice]
+    arr_perturbed[indices] = (arr_perturbed + noise)[indices]
+
     return arr_perturbed
 
 
-def uniform_sampling(arr: np.array, perturb_radius: float = 0.02, **kwargs) -> np.array:
+def uniform_noise(
+    arr: np.array,
+    indices: Union[int, Sequence[int], Tuple[np.array]],
+    indexed_axes: Sequence[int],
+    perturb_radius: float = 0.02,
+    **kwargs,
+) -> np.array:
     """Add noise to input as sampled uniformly random from L_infiniy ball with a radius."""
+
+    indices = expand_indices(arr, indices, indexed_axes)
     noise = np.random.uniform(low=-perturb_radius, high=perturb_radius, size=arr.shape)
```

Comment:
can we add function arguments to have options to set both low and high?

i.e., 
lower_bound: float = 0.02,
upper_bound: float = None,

if upper_bound is None:
    noise = np.random.uniform(low=-lower_bound, high=lower_bound, size=arr.shape)
else:
    # TODO make an assert saying that upper_bound > lower_bound
    noise = np.random.uniform(low=lower_bound, high=upper_bound, size=arr.shape)

also update where uniform_noise is the default peturb_func, like in robustness metric

Reply:
fantastic thanks